docs/config.md
# Configuration File

Some general variables that can be configured in the `wallust.toml` file.

**All variables are optional**, as wallust can work without a config
file and choose default implementations.

Keep in mind that backend, palette and colorspace share possible values
with the cli.

## backend
Allows you to choose which method to use in order to parse the image.

> **full**
>
>    Read and return the whole image pixels more precision slower.
>
> **resized**
>
>    Resizes the image before parsing mantaining it s aspect ratio.
>
> **wal**
>
>    Uses image magick convert to generate the colors like pywal.
>
> **thumb**
>
>    Faster algo hardcoded to x no ratio respected.
>
> **fastresize**
>
>    A much faster resize algo that uses SIMD. For some reason it fails
>     on some images where resized doesn t for this reason it doesn t
>     replace but rather it s a new option.
>
> **kmeans**
>
>    Kmeans is an algo that divides and picks pixels all around the
>     image. Requires more tweaking and more in depth testing but for
>     the most part it just werks.

## color_space

:   What colorspace to use to gather the most prominent colors.

> **lab**
>
> :   Uses Cie L a b color space. *(mixed)*
>
> **lch**
>
> :   CIE Lch, you can understand this color space like LAB but with
>     chrome and hue added, which Could help when sorting. *(mixed)*

> There is one variant: **mixed**, which mixes colors when collecting
> them into a histogram.

## threshold

Wallust automatically uses the **best threshold** (by my testings) if this variable isn't defined.

>    If you really want to define this variable, keep in mind the
>    following. Thershold is the **difference between similar colors** ,
>    used inside the colorspace. While each colorspace may have different
>    results with different thresholds, meaning **you should try which
>    one works for you best** , an "overall" and general table looks
>    like this:

| Number | Description |
|--------|-------------|
1 | Not Perceptible by human eyes.
1 \- 2|Perceptible through close observation.
2 \- 10|Perceptible at a glance.
11 \- 49|Colors are more similar than opposite.
100|Colors are exact opposite.

## palette
Uses the colors gathered from `color_space` in a way that makes sense,
resulting in a scheme palette.

> **dark**
>
>    Dark colors dark background and light contrast. *(16, comp,
>     comp16)*
>
> **harddark**
>
>    Same as dark with hard hue colors. *(16, comp, comp16)*
>
> **light**
>
>    Light bg dark fg. *(16, comp, comp16)*
>
> **softdark**
>
>    Variant of softlight uses the lightest colors and a dark
>     background could be interpreted as dark inversed. *(16, comp,
>     comp16)*
>
> **softlight**
>
>    Light with soft pastel colors counterpart of harddark. *(16, comp,
>     comp16)*


### palette variants
There are some variants to the principal palettes schemes which you can use by
appending the variant to the name e.g. '`dark16`', '`lightcomp`', '`harddarkcomp16`'
and so on, each palette indicates, in parenthesis, which variants are
avaliable. A description of the overall variants:

> **16**
>
>    Makes shades of colors, creating the ilusion of "16 different
>     colors".
>
> **comp**
>
>    Stands for Complementary and completly changes the palette to
>     it's complementary counterpart.
>
> **comp16**
>
>    Complementary palette with 16 shades.

## check_constrast
Ensures a "readable contrast".
Should only be enabled when you notice an unreadable contrast frequently
happening with your images. The reference color for the contrast is the
background color. (default: **disabled**)


## fallback_generator
This field chooses a method to use when the gathered colors aren't enough:

    **interpolation**

    :   (default) Tries to pick two colors and built gradients over them
        **complementary** Uses the complementary colors of two colors,
        or more (if needed), colors.

## saturation
Color saturation, usually something higher than 50 increases the saturation and
below decreases it (on a scheme with strong and vivid colors).

_Possible values:_ 1 - 100 (default: **disabled**)

## alpha
Alpha value for templating (default: **100**)

# Templates

Templates are optional and defined inside the `[templates]` header. Here it's
recommended to use single quotes (`'`) instead of double quotes (`"`) since the
first one, by the toml format, ignores backslashes (`\`) as escape codes,
allowing you to define Widows like paths (e.g. `'C:\Users\Desktop\'`)

## template
A relative path that points to a file where wallust.toml is located, usually at
`~/.config/wallust/`. This file can also be a directory, which will be
templated **non-recursively**.

## target
Absolute path in which to place a file with generated templated values. This
field CAN expand the \`\~\` as the \`\$HOME\` enviromental variable. If
*template* is a directory, this **must** correspond and be one.

## pywal _(optional)_
Indicates to treat **template** as a pywal template, using `{variable}` syntax. (default: **false**)

# Example
Below is a simple example exahusting all possible cases. All the format
is correct:

```toml
backend = "full"
color_space = "lab"
threshold = "20"
palette = "softdark"
check_contrast = true

[templates]
# dunst templates
dunst.template = "dunstrc.monitor"
dunst.target = "~/.config/dunst/dunstrc"

# one liner for zathura
zathura = { template = 'zath', target = '~/.config/zathura/zathurarc' }

# even a shorter way
glava = { src = 'glava.glsl', target = '~/.config/glava/rc.glsl' }

# or splited, like dunst
res.src = "xres"
res.dst = "~/.config/Xresources"

# old times, good times (here I put old pywal templates)
# Also, note that the name doesn't matters (as long as it isn't already defined)
# NOTE that both src and dst are directories
pywal = { src = "templates/", dst = '~/.cache/wal/', pywal = true }
```


src/config.rs
//! Config related stuff, like parsing the config file and writing templates defined on it
use std::collections::HashMap;
use std::path::Path;
use std::path::PathBuf;
use std::fs::read_to_string;

use crate::args::WallustArgs;
use crate::colors::Colors;
use crate::template;
use crate::template::TemplateFields;

use anyhow::{Result, Context};
use owo_colors::{AnsiColors, OwoColorize};
use serde::Deserialize;

/// Representation of the toml config file `wallust.toml`
///
/// Maybe divide into `Internal` and `TomlConfig`.
#[derive(Debug, Deserialize, Default)]
#[cfg_attr(feature = "doc" , derive(documented::Documented, documented::DocumentedFields))]
pub struct Config {
    /// threshold to use to differentiate colors
    #[serde(default)]
    #[serde(deserialize_with = "validate_threshold")]
    pub threshold: Option<u8>,
    /// Which backend to use, see backends.rs
    #[serde(rename = "backend")]
    pub backend_de: Option<crate::backends::Backend>,
    /// Which palette to use, see palettes.rs
    #[serde(rename = "palette")]
    pub palette_de: Option<crate::palettes::Palette>,
    /// Which colorspace to use, see colorspaces.rs
    #[serde(rename = "color_space")]
    pub color_space_de: Option<crate::colorspaces::ColorSpace>,
    /// Optional alpha value
    pub alpha: Option<u8>,
    /// This flags ensures good contrast between images, by doing some w3m calculations.
    /// However it isn't required and should only be turn on when you notice bad contrast between many images.
    pub check_contrast: Option<bool>,
    /// Maybe the user requires more vivid colors
    pub saturation: Option<u8>,
    /// How to 'generate' colors when there aren't enough colors to create the `palette`.
    /// This appears as "Artificially generating colors.." in cli
    pub fallback_generator: Option<crate::colorspaces::FallbackGenerator>,
    /// templates: a new way of defining templates, giving the ability of naming stuff.
    // [templates]
    // dunst.src = 'C:\long\path'
    // dunst.dst = '~/.config/dunst'
    // zathura = { src = 'zathura.rc', dst = '~/.config/zathura' }
    pub templates: Option<HashMap<String, Fields>>,
    /// Allows to change the path of the templates
    ///  default: "~/.config/wallust/templates/"
    //pub templates_dir: PathBuf,

    #[deprecated]
    /// TOML: array of tables for "template" and "target"
    /// This is here only for `wallust migrate`
    pub entry: Option<Vec<Entries>>,

    /// Config directory (wallust/) path
    #[serde(skip)]
    pub dir: PathBuf,

    /// Config file (wallust.toml) path
    #[serde(skip)]
    pub file: PathBuf,

    /* TRUE VALUES , used internally */

    /// True threshold gathered from threshold.
    #[serde(skip)]
    pub true_th: u8,

    #[serde(skip)]
    pub backend: crate::backends::Backend,

    #[serde(skip)]
    pub color_space: crate::colorspaces::ColorSpace,

    #[serde(skip)]
    pub palette: crate::palettes::Palette,
}

/// An entry within the config file, toml table
/// ref: <https://toml.io/en/v1.0.0#array-of-tables>
#[derive(Debug, Deserialize, Clone)]
pub struct Fields {
    /// A file inside `~/.config/wallust/`, which is used for templating
    #[serde(alias = "src")]
    pub template: String,
    /// Where to write the template
    #[serde(alias = "dst")]
    pub target: String,
    /// Allows pywal template spec compatibility (disabled by default)
    pub pywal: Option<bool>,
    // If 'src' is a directory, 'dst' SHOULD also be one.
    // This flag allows for 'src', when a dir, to be templated recursively
    // If 'src' is a file, this has no effect.
    //TODO implement recursive behaviour
    //pub recursive: Option<bool>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Entries {
    /// A file inside `~/.config/wallust/`, which is used for templating
    pub template: String,
    /// Where to write the template
    pub target: String,
    /// Allows pywal template spec compatibility (disabled by default)
    pub new_engine: Option<bool>,
}

/// How to populate `wallpaper` template value:
/// 1. With `wallust theme rose-pine`, it will use the name of the theme in use. (e.g. `rose-pine`)
/// 2. With `wallust cs scheme.json`, it will use the absolute path of the file used. (e.g. `/home/user/scheme.json`)
/// 3. Normal behaviour with `wallust run image.png`, it will use the wallpaper absolute path. (e.g. `/home/user/image.png`)
pub enum WalStr<'a> {
    Path(&'a Path),
    Theme(&'a str),
}

/// v3.md link
pub const V3: &str = "<https://codeberg.org/explosion-mental/wallust/src/tag/3.0.0-alpha/v3.md>";

impl Config {
    /// Constructs [`Config`] by reading the config file
    pub fn new(original_config_path: &PathBuf, config_file: Option<&Path>, config_dir: Option<&Path>, no_config: bool) -> Result<Config> {

        // check config file or generate one if not one isn't found
        let custom = config_file;

        // true -> uses original_config_path
        // false -> uses a custom path
        let mut is_original = true;

        // check config dir
        let config = match config_dir {
            Some(path) => { //only in this case, the config dir is altered
                is_original = false;
                path
            },
            None => original_config_path,
        };

        // init `.config/wallust/wallust.toml`
        let join_dir = if is_original { "wallust" } else { "" };

        let config_dir = config.join(join_dir);
        let def_conf = config_dir.join("wallust.toml");

        // is the user using `--config-path`
        let (config, _default_path) = match custom {
            None => (def_conf, true),
            Some(s) => (s.to_owned(), false),
        };

        let mut ret = if !config.exists() || no_config {
            Config::default()
        } else {
            let s = || format!("Failed to read file {}:\nIf you are switching from v2 to v3, use `wallust migrate`.\nMake sure to read {V3} as well.", config.display());
            toml::from_str(
                &read_to_string(&config).with_context(s)?
            ).with_context(s)?
        };

        ret.dir = config_dir;
        ret.file = config.to_path_buf();
        ret.true_th = 0;

        // defined or defaults.
        ret.backend = ret.backend_de.unwrap_or_default();
        ret.color_space = ret.color_space_de.unwrap_or_default();
        ret.palette = ret.palette_de.unwrap_or_default();

        //println!("{:#?}", ret);

        Ok(ret)
    }

    pub fn print(&self) {
        let empty = String::new();
        let k = if self.check_contrast.unwrap_or(false) {
            format!("\n[{}] {}: Doing extra calculations to ensure a good contrast",
                "I".blue().bold(),
                "contrast".magenta().bold()
                )
        } else { empty.clone() };

        let sat = if let Some(s) = self.saturation {
            format!("\n[{}] {}: Adding saturation to existing palette by {s}%",
                "I".blue().bold(),
                "saturation".magenta().bold()
                )
        } else { empty };

        let th = match self.threshold {
            Some(s) => format!("Using a threshold of {s} in between colors."),
            None => format!("Not defined, using {} default thresholds.", "best".bold()),
        };

        println!(
"[{i}] {back_f}: Using {back} backend parser
[{i}] {th_f}: {th}
[{i}] {cs_f}: Using {cs} colorspace variation
[{i}] {palette_f}: Using {palette} palette{k}{sat}",
            back     = self.backend.bold().color(self.backend.col()),
            palette  = self.palette.bold().color(self.palette.col()),
            cs       = self.color_space.bold().color(self.color_space.col()),
            i        = "I".blue().bold(),
            back_f   = "image parser".magenta().bold(),
            th_f     = "threshold".magenta().bold(),
            palette_f = "scheme palette".magenta().bold(),
            cs_f     = "colorspace".magenta().bold(),
        );
    }

    /// Writes templates defined in the config file (if any)
    /// Should print a warning if you are using the old `[[entry]]` syntax (since it's going to be deprecated in v3).
    pub fn write_entry(&self, wal_str: &WalStr, colors: &Colors, quiet: bool) -> Result<()> {
        let init = format!("[{info}] {t}: ", info = "I".blue().bold(), t = "templates".magenta().bold());

        let templates_header = match &self.templates {
            Some(s) => {
                if ! quiet { println!("{init}Writing templates.."); }
                s
            },
            None => {
                if ! quiet { println!("{init}No templates found"); }
                return Ok(())
            },
        };

        // check if themes exist, if it does we are using the `theme` subcommand,
        // which means there is not image path, so use the theme name as for the `wallpaper` value
        let image_path = match wal_str {
            // use the theme name otherwise
            WalStr::Theme(s) => s.to_string(),
            // make sure to display the absolute path of the wallpaper
            WalStr::Path(p) => dunce::canonicalize(p).expect("PATH EXIST, validation from clap").display().to_string(),
        };

        let values = TemplateFields {
            alpha: self.alpha.unwrap_or(100),
            backend: &self.backend,
            colorspace: &self.color_space,
            palette: &self.palette,
            image_path: &image_path,
            colors,
        };

        template::write_template(&self.dir, templates_header, &values, quiet)
    }

    /// if the user provides this values in the cli, overwrite the [`Config`] configuration
    pub fn customs_cli(&mut self, cli: &WallustArgs) {
        if let Some(b) = cli.backend {
            self.backend = b;
        }

        if let Some(col) = cli.colorspace {
            self.color_space = col;
        }

        if let Some(f) = cli.palette {
            self.palette = f;
        }

        if let Some(t) = cli.threshold {
            self.threshold = Some(t as u8); //t is [1..=100]
        }

        if let Some(a) = cli.alpha {
            self.alpha = Some(a as u8);
        }

        if cli.check_contrast {
            self.check_contrast = Some(cli.check_contrast);
        }

        if let Some(sat) = cli.saturation {
            self.saturation = Some(sat as u8);
        }

        if let Some(g) = cli.fallback_generator {
            self.fallback_generator = Some(g);
        }
    }

    /// thershold color for owo_colors
    pub fn threshold_col(&self) -> AnsiColors {
        match self.true_th {
            1 => AnsiColors::Yellow,
            2 => AnsiColors::Cyan,
            3..=10 => AnsiColors::Green,
            11..=49 => AnsiColors::Blue,
            50..=100 => AnsiColors::Red,
            _ => AnsiColors::Red,
        }
    }
}

impl std::fmt::Display for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
            let sp = "    ";

            let temps = if let Some(e) = &self.templates {
                let mut s = String::new();
                for i in e {
                    let pywal = if let Some(s) = i.1.pywal {
                        format!("{sp}{sp}pywal = {s}\n")
                    } else {
                        String::new()
                    };

                    let name = i.0;

                    s.push_str(
                        &format!("{sp}{name}\n{sp}{sp}template = {}\n{sp}{sp}target   = {}\n{pywal}",
                                i.1.template, i.1.target)
                        );
                }
                s.trim_end().to_owned()
            } else {
                String::new()
            };

            let templates = if temps.is_empty() {
                "No entries found.".into()
            } else {
                temps
            };

            write!(f, "\
Config directory: {dir}
Config file: {file}
Configuration options:
    backend        = {b}
    color_space    = {c}
    threshold      = {t:?}
    palette        = {f}
    check_contrast = {con:?}
    saturation     = {sat:?}
    alpha          = {a:?}
Templates:
{templates}",
            b = self.backend,
            c = self.color_space,
            t = self.threshold,
            f = self.palette,
            con = self.check_contrast,
            sat = self.saturation,
            a = self.alpha,
            dir = self.dir.display(),
            file = self.file.display(),
            )
    }
}

fn validate_threshold<'de, D>(d: D) -> Result<Option<u8>, D::Error>
    where D: serde::de::Deserializer<'de>
{
    use serde::de;


    let value = Option::deserialize(d)?;
    let value = match value {
        Some(s) => s,
        None => return Ok(None),
    };

    if value <= 100 { return Ok(Some(value)); }

    Err(de::Error::invalid_value(de::Unexpected::Unsigned(value as u64), &"a value between 0 and 100."))
}
